### PR TITLE
Fix thrust compilation in MSVC 14 for CUDA 11.2

### PR DESCRIPTION
**Pull request by emcastillo**
The headers changed to have `constexpr`

### CODE REVIEW

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
seems hipcc didn't like the linebreak ...

**Comment by leofang** [comment]
I can take a look.

**Comment by leofang** [comment]
I think using `THRUST_OPTIONAL_CPP11_CONSTEXPR` from `thrust/optional.h` is a lot cleaner, but it is not preset until CUDA 10.1 😅

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
AMD build test passed.

**Comment by leofang** [comment]
@kmaehashi This is ready.

File: cupy/cuda/cupy_thrust.cu [hunk 1]
```diff
@@ -126,31 +126,47 @@ __host__ __device__ __forceinline__ bool _cmp_less(const T& lhs, const T& rhs) {
 
 // specialize thrust::less for single complex
 template <>
-__host__ __device__ __forceinline__ bool less<complex<float>>::operator() (
+__host__ __device__ __forceinline__
+#if (__CUDACC_VER_MAJOR__ == 11 && __CUDACC_VER_MINOR__ == 2)
```
Comment (leofang) [inline cupy/cuda/cupy_thrust.cu:130]: This should've been 
```C++
#if (__CUDACC_VER_MAJOR__ == 11 && __CUDACC_VER_MINOR__ >= 2)
```
I can send a PR after code freeze.

Comment (leofang) [reply cupy/cuda/cupy_thrust.cu:130]: Better:
```
#if (__CUDACC_VER_MAJOR__ >11 || (__CUDACC_VER_MAJOR__ == 11 && __CUDACC_VER_MINOR__ >= 2))
```

Comment (emcastillo) [reply cupy/cuda/cupy_thrust.cu]: Changed, thanks!

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by kmaehashi** [comment]
pfnCI, test this please.

**Comment by kmaehashi** [comment]
pfnCI, test this please.

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Jenkins failures are unrelated so let's not trigger CIs again

**Comment by kmaehashi** [comment]
LGTM! Test failures are not related to this PR.

**Pull request merged**